Sources/Bookmarks/FaviconsFetcher/FaviconsFetchOperation.swift

```diff
@@ -0,0 +1,376 @@
+        } catch let error as FaviconFetchError {
+            switch error {
+            case .connectionError:
+                return []
+            case .requestError:
+                return bookmarkIDs
+            }
+        }
+
+    }
+
+    private func fetchAndStoreFavicon(for url: URL) async throws {
+        let fetchResult: (Data?, URL?)
+        do {
+            fetchResult = try await fetcher.fetchFavicon(for: url)
+        } catch {
+            let nsError = error as NSError
+            // if user is offline, we want to retry later
+            let temporaryErrorCodes = [NSURLErrorNotConnectedToInternet, NSURLErrorTimedOut, NSURLErrorCancelled]
+            if nsError.domain == NSURLErrorDomain, temporaryErrorCodes.contains(nsError.code) {
+                throw FaviconFetchError.connectionError
+            }
+            throw FaviconFetchError.requestError
+        }
+
+        do {
+            let (imageData, imageURL) = fetchResult
+            if let imageData {
+                os_log(.debug, log: log, "Favicon found for %{public}s", url.absoluteString)
+                try await faviconStore.storeFavicon(imageData, with: imageURL, for: url)
+            } else {
+                os_log(.debug, log: log, "Favicon not found for %{public}s", url.absoluteString)
+            }
+
+            try checkCancellation()
+        } catch is CancellationError {
+            os_log(.debug, log: log, "Favicon fetching cancelled")
+            throw CancellationError()
+        } catch {
+            os_log(.debug, log: log, "Error storing favicon for %{public}s: %{public}s", url.absoluteString, error.localizedDescription)
+            throw error
+        }
+    }
+
+    private func checkCancellation() throws {
+        if isCancelled {
+            throw CancellationError()
+        }
+    }
+
+    private func mapBookmarkDomainsToUUIDs(for uuids: any Sequence & CVarArg) -> BookmarkDomains {
+        let request = BookmarkEntity.fetchRequest()
+        request.predicate = NSPredicate(
+            format: "%K IN %@ AND %K == NO AND %K == NO",
+            #keyPath(BookmarkEntity.uuid), uuids,
+            #keyPath(BookmarkEntity.isFolder),
+            #keyPath(BookmarkEntity.isPendingDeletion)
+        )
+        request.propertiesToFetch = [#keyPath(BookmarkEntity.url)]
+        request.relationshipKeyPathsForPrefetching = [#keyPath(BookmarkEntity.favoriteFolders), #keyPath(BookmarkEntity.parent)]
+
+        let context = database.makeContext(concurrencyType: .privateQueueConcurrencyType)
+        var bookmarkDomains: BookmarkDomains!
+        context.performAndWait {
+            let bookmarks = (try? context.fetch(request)) ?? []
+            bookmarkDomains = .init(bookmarks: bookmarks)
+        }
+        return bookmarkDomains
+    }
+
+    private var log: OSLog {
+        getLog()
+    }
+    private let getLog: () -> OSLog
+
+    // MARK: - Overrides
+
+    override var isAsynchronous: Bool { true }
```

Comment:
Do we need this? You seem to be always starting operation through the queue?

Reply:
I think we need this in order for the operation's `main()` function to not report finished state after exiting, but fait for isFinished to be set via an async Task that's run there. It's the same for SyncOperation.

Reply:
I don't think it hurts, but I belive this value is ignored by OperationQueue - as it always starts the operation on a queue. Nonetheless we can keep it. :)